auth/common.hh

```diff
@@ -95,12 +96,61 @@ future<> announce_mutations_with_batching(
         std::function<mutations_generator(api::timestamp_type& t)> gen,
         seastar::abort_source* as);
 
-// Execute update query via group0 mechanism, mutations will be applied on all nodes.
+// FIXME(mmal): remove
 future<> announce_mutations(
         cql3::query_processor& qp,
         ::service::raft_group0_client& group0_client,
         const sstring query_string,
         std::vector<data_value_or_unset> values,
         seastar::abort_source* as);
 
+// mutations_collector is used to gather mutations which are side effects
+// of functions execution. They need to be announced under single guard
+// for atomicity. As functions which produce mutations may embed each other
+// we need to decouple announcing step to a common external place here.
+// It also supports generator callbacks to avoid holding too many mutations
+// in memory as number of mutations may depend on user's data.
+// There is no guarantee about ordering of mutations.
+// If size of mutations is too big for raft too handle under single guard
+// atomicity guarantee may be violated.
+class mutations_collector {
+public:
+    using generator_func = std::function<mutations_generator(api::timestamp_type& t)>;
+private:
+    std::vector<mutation> _muts;
+    std::vector<generator_func> _generators;
+    std::optional<::service::group0_guard> _guard;
+public:
+    mutations_collector(::service::group0_guard g) : _guard(std::move(g)) {}
+    // This constructor is used to handle both legacy and current paths
+    // as there is no guard for legacy and the whole class is only passed
+    // through to simplify the code.
+    mutations_collector(std::optional<::service::group0_guard> g) {
+        if (g) {
+            _guard = std::move(*g);
+        }
+    };
+    ~mutations_collector();
+
+    mutations_collector(mutations_collector&) = delete;
+    mutations_collector(mutations_collector&&) = default;
+
+    // gets timestamp which should be used when building mutations
+    api::timestamp_type write_timestamp() const;
+
+    void add_mutation(mutation m);
+    void add_mutations(std::vector<mutation> ms);
+    void add_generator(generator_func f);
+    // announce commits all data, must be called only once. Nop if there was no guard provided.
+    future<> announce(::service::raft_group0_client& group0_client, seastar::abort_source& as);
```

Comment:
that's because add_mutation doesn't do any splitting, as you pointed out it would break, and I don't expect it to be needed with 1-4 mutations it's being used for. 

Perhaps std::vector<> overload is a bit misleading it suggests that there can be a lot of mutations, but that's only because other internal functions return vector even despite creating 1 mutation.

Reply:
Right, but the current interface allows adding both generators and mutations, and mutations are appended as a generator to the end of the generator list:

```c++
    if (_muts.size() > 0) {
        _generators.emplace_back([this] (api::timestamp_type&) -> mutations_generator {
            for (auto& mut: _muts) {
                co_yield std::move(mut);
            }
            _muts.clear();
        });
    }
```

So, if one of the generators added by the user creates a lot of mutations that overflow a single group0 operation, the last generator that yields mutations added by the user will return mutations with the wrong timestamps.

Perhaps the generator list should be reversed before being merged, or the last generator should be prepended to `_generators` instead of being appended at the end.

Reply:
Does generators even used in the series? I did not see it. May be we can drop it for now?

Reply:
it's a draft, I have usage with generator in this PR to do. Didn't do yet as wanted to clarify the approach first, it's a lot of refactor to move everything.